### ISSUE TITLE
More `kubectl create` sub-commands

### ISSUE THREAD

**Issue by janetkuo** [kubernetes/kubernetes#25382]
Adding more `kubectl create` sub-commands make resource creation much easier. For example, 

```
kubectl create pods
kubectl create rc
kubectl create services
kubectl create deployments
...
```

with `--edit` (#18064) it will be very convenient

Original issue: #10495

@kubernetes/kubectl

**Comment by kargakis** [comment]
Names should be singular. Also does it make sense to have create subcommands for all kinds of resources? For example, why have `kubectl create rc` when there is `kubectl create deployment`?

**Comment by AdoHe** [comment]
I agree that names should be singular. For me, have create subcommands for all kinds resources does make sense, I like that.

**Comment by deads2k** [comment]
> For example, why have kubectl create rc when there is kubectl create deployment?

Even though we may prefer one kind of resource to another, I think that avoiding a create subcommand for particular resources will only frustrate users who want to create it, not guide them to creating something different.

**Comment by deads2k** [comment]
I think that `kubectl create service` is a great idea, but I suspect there are a couple flavors lurking there.  There's the external IP case, node port case, load balancer case, and headless case.  I don't know that we care about all of those, but I think we should make an explicit choice one way or the other.

If I'm creating a NodePort service, having a `--headless` (or whatever we'd call it) flag doesn't make much sense.

If i'm creating a ClusterIP service, do we think that specifying a specific clusterIP is a common enough use-case that we need a flag?  I think I'd be against exposing it just because its a pretty advanced thing to do that requires global knowledge.

**Comment by asalkeld** [comment]
I can implement this if no one else has started.

**Comment by asalkeld** [comment]
Some ideas for Services and Deployments

**Services**
Firstly ignore more advanced features as @deads2k suggested:
- clusterIP
-  (within the port definition) name and nodePort
- loadbalancerIP (not always available)

create service clusterIP <name>  [selector, ...] [port, ..] --externalIP=1.2.3.4
create service NodePort <name>  [selector, ...] [port, ..]
create service LoadBalancer <name> [selector, ...] [port, ..] --ingressIP=1.2.3.4

where:
 --selector="key[op]value" (can be specified multiple times)
 --port="TCP,80,937" (can be specified multiple times)
 --label="key=value" (can be specified multiple times) 

example:
```
create service clusterIP redis-master --label="app=redis" --label="role=master" --label="tier=backend" --port="TCP,6379,6379" --selector="app==redis" --selector="role==master" --selector="tier==backend"
```

**Deployments**

assumptions:
- "name" goes into (metadata/name, spec/template/metadata/labels/deployment)

create deployment <name> <replicas> <list of containers[name, image, port]>

where:
--label="key=value" (can be specified multiple times) goes into  spec/template/metadata/labels/<key>
 --container="<name>,<image>,<port>,<port>"

example:
```
create deployment redis-master 1 --container="master,gcr.io/google_containers/redis:e2e,6379" --label="app=redis" --label="role=master" --label="tier=backend"
```

**Questions**
- does that seem reasonable?
- is the --port and --container comma delimited list OK?

**Comment by deads2k** [comment]
I like the proposal for the different types of services.  I'm a little unsure of `--externalIP`.  It takes some advanced configuration to set that up to work right.  Do you find people are using that instead of ingress?

I think that we might want a `--headless` flag for the clusterip service.  It's an immutable field that is used to by things that want to locate endpoints in DNS.

Do you prefer `--port=TCP,4321 --port=UDP=1234` to `--tcp=4321 --udp=1234`?  Seems like the latter choice keeps a user from having to specify strings in a particular format.

Our guiding principle on `kubectl create <subcommand>` is to have as few flags as possible to create a valid object (https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#create-commands).  That means that things like `--label` and `<replicas>` are probably out for an initial implementation.  To handle that use-case I think I'd like to be able to do something like `kubectl create service cluster-ip -o yaml --dry-run | kubectl label --local-file - app=redis -o yaml | kubectl create -f -` and `kubectl create deployment -o yaml --dry-run | kubectl scale --local-file - 3 | kubectl create -f -`.

**Comment by deads2k** [comment]
More feedback about `--container` on deployments.  Is there a strong case for needing to specify the name or could you choose a reasonable format based on the image name?  Similarly, could ports be lifted from the image metadata if available or left blank otherwise?  Seems like the majority of users aren't likely to be changing the default ports on a simple command like this.

**Comment by asalkeld** [comment]
Great feedback @deads2k , definitely less options on a command set that is supposed to be easy to use is the way to go (we can always add an option later if needed).

1. I'll drop the labels and only add the name as a label (so if you create the service and deployment with the same name it should work).
2. I'll drop --externalIP
3. I'll add --headless
4. the --container=image seems doable to me..
5.  the port object has a targetPort, that seemed important to me, but maybe it's not needed here? I am honestly not sure just how often it's used. If we don't need it then --tcp=x seems the nicest option IMHO.

**Comment by deads2k** [comment]
> the port object has a targetPort, that seemed important to me, but maybe it's not needed here? I am honestly not sure just how often it's used. If we don't need it then --tcp=x seems the nicest option IMHO.

That's a good point.  Maybe allow both `tcp=1234` and `tcp=1234=4321`?  I think that people just starting aren't likely to skew them, but something like that syntax keeps the easy case easy and keeps the slightly more complicated case reasonable.

**Comment by smarterclayton** [comment]
If Docker has already established a convention with port mapping syntax
(HOST:CONTAINER) we should lean towards using that rather than inventing a
new format.

**Comment by deads2k** [comment]
> If Docker has already established a convention with port mapping syntax
> (HOST:CONTAINER) we should lean towards using that rather than inventing a
> new format.

Sure, no objection to a different delimiter that isn't a comma.

**Comment by derekwaynecarr** [comment]
I don't think create deployment needs to optimize around pod templates with multiple containers, so generally agree with @deads2k previous feedback as well.  I am not sure I think we need sub commands for create service based on service type versus just `kubectl create service --type=clusterip` similar to `kubectl expose`.  Seems like `kubectl create service` should be everything that is unrelated to the `TYPE NAME` parts of `kubectl expose`.

**Comment by deads2k** [comment]
> kubectl create service --type=clusterip similar to kubectl expose

This actually highlights the problem that we need to solve in `kubectl create`.  `kubectl expose` has multiple conflicting flags that are only valid in certain combinations based on other flags.  In the case of the example you just gave, adding a `--type` to a `kubectl create service` will put us in a position where we again have flags that don't make sense in all contexts.  For instance, what would ` --headeless` mean if `--type=NodePort`?  It's an invalid combination that we can prevent given specific flavors.

If someone is using this command, they either know which kind of service they want or they want a vanilla clusterip one.  I could see mapping `kubectl create service` to `kubectl create service clusterip`, but I don't think that a `--type` argument will address the problems we're having with other multi-flavor commands that are unable to provide meaningful help or predictable results.

**Comment by derekwaynecarr** [comment]
@deads2k - I can see that.   I see that I was somewhat contradictory in my thinking.  Either way, I like the idea of supporting the bare minimum amount of flags possible, and so I was reconciling that with our ability to change the `--type` of a service post creation (which I think was a mistake in the API).

**Comment by pwittrock** [comment]
Some state in #34731

**Comment by pwittrock** [comment]
@apelisse

**Comment by AdoHe** [comment]
@pwittrock should we use this issue as tracking `more create sub-commands`?  you know we plan to add more sub-commands in Q1

**Comment by pwittrock** [comment]
Yeah we can use this for now.

**Comment by adohe** [comment]
I have open #44936 to move all create commands to a separate package, ptal.

**Comment by soltysh** [comment]
@ercliou sure go ahead.

**Comment by almiche** [comment]
👋 I'm wondering if anyone thinks that `kubectl create pods --image=busybox:tag` is worthwhile adding in. I would like to take a stab at it, so far looking at the source code I haven't seen this new sub-command added in.

**Comment by modevops** [comment]
I am frustrated with kubectl create because there are very little options.  Why when kubectl create deployments  was created did not have the same features added and kubectl run.  --replicas at the very least should have been added. I found the kubectl create command very limited.  It would be nice to be able to do more the create a deployment and add an image.  Another good options to set values for the deployments like spec.containers[0].volumeMounts or spec.containers[0].commands  It would also be great create pv and pvc. Right now to me the command set very limited.
